File: services/spring/order_service/src/main/java/com/smartfulfillment/order_service/client/WarehouseClient.java
package com.smartfulfillment.order_service.client;

import com.smartfulfillment.order_service.dto.StockDTO;
import org.springframework.cloud.openfeign.FeignClient;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;

import java.util.List;
import java.util.Map;
import java.util.UUID;

@FeignClient(name = "warehouse-service", url = "${warehouse.service.url:http://localhost:8084}")
public class WarehouseClient {
    @GetMapping("/stock/{productId")
    List<StockDTO> getStockByProduct(@PathVariable("productId")UUID productId);

    @PostMapping("/warehouse/{warehouseId}/stock")
    void updateStock(@PathVariable("warehouseId") UUID warehouseId, @RequestBody Map<String, Object> payload);
}

File: services/go/order-twin/internal/clients/warehouse_client.go
package clients

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/google/uuid"
	"github.com/smartfulfillment/order-twin/internal/config"
)

// StockDTO matches the Java StockDTO exactly
type StockDTO struct {
	WarehouseID   uuid.UUID `json:"warehouse_id"`
	WarehouseName string    `json:"warehouse_name"`
	Location      string    `json:"location"`
	Quantity      int       `json:"quantity"`
	Latitude      float64   `json:"latitude"`
	Longitude     float64   `json:"longitude"`
}

// GetStockByProduct mimics GET /stock/{productId}
func GetStockByProduct(productID uuid.UUID) ([]StockDTO, error) {
	cfg := config.LoadConfig()
	url := fmt.Sprintf("%s/stock/%s", cfg.WarehouseServiceURL, productID.String())

	// Use the package-level sharedClient
	resp, err := sharedClient.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("warehouse service returned status: %d", resp.StatusCode)
	}

	var stocks []StockDTO
	if err := json.NewDecoder(resp.Body).Decode(&stocks); err != nil {
		return nil, err
	}

	return stocks, nil
}

// UpdateStock mimics POST /warehouses/{warehouseId}/stock
func UpdateStock(warehouseID uuid.UUID, payload map[string]interface{}) error {
	cfg := config.LoadConfig()
	url := fmt.Sprintf("%s/warehouses/%s/stock", cfg.WarehouseServiceURL, warehouseID.String())

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	// Use the package-level sharedClient
	resp, err := sharedClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return fmt.Errorf("warehouse service failed to update stock, status: %d", resp.StatusCode)
	}

	return nil
}
